### ISSUE TITLE
The fields `created_by` of PageUser, PageGroup and `user`/`group` of PagePermission can cause unwanted user/group deletion

### ISSUE THREAD

**Issue by benzkji** [django-cms/django-cms#7225]
## Description

As all of these fields are defined as `on_delete=models.CASCADE`. That means, if I delete User A, that has previously created a `PageUser` B, the user B will be deleted as well. I never was really sure what these are anyway... see also #5517 .

## Steps to reproduce

Create User A. Log in with A, create PageUser B. Delete User A. User B is gone as well.

## Expected behaviour

User B should not be deleted. Instead, one could show (deleted User), and define the fields as `on_delete=models.SET_NULL`. Or remove the fields completly.

## Actual behaviour

User B is gone/deleted.   

Also, discovered that normal new Users that are created trigger the creation of a `PageUser`, if `CMS_PERMISSION` is true. This can cause confusion, and makes it almost impossible to delete useres on such a system.

## Additional information (CMS/Python/Django versions)

for discussion since 2016 #5517 

## Do you want to help fix this issue?

* [ ] Yes, I want to help fix this issue and I will join #workgroup-pr-review on [Slack](https://www.django-cms.org/slack) to confirm with the community that a PR is welcome. 
* [x] No, I only want to report the issue.

**Comment by benzkji** [comment]
anyone? this is a candidate to shot onesels in the foot...

**Comment by marksweb** [comment]
@benzkji this needs raising in a tech committee call I think.

**Comment by benzkji** [comment]
yep, think so. removing `created_by` would be no big deal (or use on_delete=models.SET_NULL).

**Comment by marksweb** [comment]
@benzkji i think the issue is that it was agreed no more migrations would go into v3 due to the impact that has on v4

**Comment by creyD** [comment]
Yes #5517 is a duplicate, and should probably be closed as such. As I added there removal is not even necessary, it would suffice to convert the PageUser to be connected to the AUTH_USER_MODEL by a one-to-one-field.

BTW: This issue is from 2015 originally and there were other ideas discussed in the previous tickets.

**Comment by bethlakshmi** [comment]
Any thoughts on this issue?  I'm getting another permutation of this when I try to upgrade Python.  On python 3.6/Django CMS 3.7 my tests all work fine.  But when I upgrade from there, I get a teardown error in my application's tests because my test cases create a User object (and a One to One custom model attached to it).  Deleting that test data through the teardown causes:
```
self = <django.db.backends.utils.CursorWrapper object at 0x7f3cd5920e50>
sql = 'SET CONSTRAINTS ALL IMMEDIATE', params = None
ignored_wrapper_args = (False, {'connection': <django.db.backends.postgresql.base.DatabaseWrapper object at 0x7f3cdc12dac0>, 'cursor': <django.db.backends.utils.CursorWrapper object at 0x7f3cd5920e50>})

    def _execute(self, sql, params, *ignored_wrapper_args):
        self.db.validate_no_broken_transaction()
        with self.db.wrap_database_errors:
            if params is None:
                # params default might be backend specific.
>               return self.cursor.execute(sql)
E               django.db.utils.IntegrityError: insert or update on table "cms_pageuser" violates foreign key constraint "cms_pageuser_created_by_id_8e9fbf83_fk_auth_user_id"
E               DETAIL:  Key (created_by_id)=(51) is not present in table "auth_user".
```

I don't work with PageUser nor do I have a custom "created_by" field.  It's just that something has shifted in the built in teardown of model objects to trigger this.  

Any idea how I can workaround it?  And/or any plans to move to a OnetoOne relationship for PageUser... which I suspect would mitigate my issue to. 

I'm blocked on upgrading either python or Django CMS without this.

**Comment by marksweb** [comment]
@fsbraun this is something we need to address I think.

**Comment by DmytroLitvinov** [comment]
Hi everyone.
Today we stacked with the same issue when we tried to delete not needed user via dj.admin and we see that it takes to delete related models which we don't want to delete.

If you need any help, I would gladly help and fix that issue and keep DjangoCMS more reliable 😊

**Comment by marksweb** [comment]
@DmytroLitvinov @fsbraun the issue here is the need for a migration. We had said we didn't want more v3 migrations because of the impact that then has on v4.

Hopefully it doesn't interfere.

**Comment by sergei-maertens** [comment]
Trying to upgrade from Django CMS 3.8 to 3.9 (and eventually 3.11) gives the exact same problem as what @bethlakshmi  is [observing](https://github.com/django-cms/django-cms/issues/7225#issuecomment-1287033808)

**Comment by brenosss** [comment]
For those looking for a quick work around the tests teardown, you can replace the `cms.middleware.user.CurrentUserMiddleware` with this:

```
class CurrentUserMiddleware(CurrentUserMiddleware):
    def __call__(self, request):
        if not settings.TEST:
            return super().__call__(request)
        response = self.get_response(request)
        return response
```

This will prevent CMS create the PageUser link after a new user is created and break the test teardown

**Comment by benzkji** [comment]
Has this been fixed in 4.1.X?

**Comment by fsbraun** [comment]
@benzkji Thanks for pinging this! No, it hasn't but needs to be fixed now. 
@vinitkumar We need to fix this before we branch off v5.

**Comment by benzkji** [comment]
@fsbraun yes, agreed.
